### ISSUE TITLE
ls: .zfs/snapshot/: Input/output error

### ISSUE THREAD

**Issue by c0xc** [zfsonlinux/zfs#8795]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| Debian
Distribution Version	| 9.5
Linux Kernel	| Linux 4.9.0-7-amd64
Architecture	| x86_64
ZFS Version	| 0.6.5.9-5
SPL Version	| 0.6.5.9-1

### Describe the problem you're observing

Trying to list the snapshots, I get i/o errors:

> ls: .zfs/snapshot/: Input/output error

Earlier, the snapshot rotation script failed because it couldn't delete some snapshots and I got these and other errors trying to access some of the snapshots:

```
# /root/bin/zfs-snapshot datapool/Temp hourly 1
cannot destroy snapshot datapool/Temp@hourly.0: dataset is busy
Error: Deleting snapshot failed!
datapool/Temp@hourly.0

# cd /datapool/Temp/.zfs/snapshot
# ls -lad hourly.*
ls: hourly.0: Input/output error
ls: hourly.1: Input/output error
dr-xr-xr-x 1 root root 0 May 19 18:33 hourly.0
dr-xr-xr-x 1 root root 0 May 19 18:33 hourly.1

$ ls -la /datapool/Temp/.zfs/snapshot/hourly.0/
ls: cannot access '/datapool/Temp/.zfs/snapshot/hourly.0/': Too many levels of symbolic links
```

The snapshot rotation script is in [c0xc/zfs-snapshot](https://github.com/c0xc/zfs-snapshot) (it's very simple).

When I saw that the snapshot names didn't match up with their mountpoints, I used `umount` to get the snapshot rotation working again:

```
# grep snapshot /proc/mounts
datapool/Temp@hourly.18 /datapool/Temp/.zfs/snapshot/hourly.40 zfs ro,relatime,xattr,noacl 0 0
datapool/Temp@hourly.14 /datapool/Temp/.zfs/snapshot/hourly.17 zfs ro,relatime,xattr,noacl 0 0
datapool/Temp@hourly.19 /datapool/Temp/.zfs/snapshot/hourly.41 zfs ro,relatime,xattr,noacl 0 0
datapool/Temp@hourly.10 /datapool/Temp/.zfs/snapshot/hourly.13 zfs ro,relatime,xattr,noacl 0 0
datapool/Temp@hourly.9 /datapool/Temp/.zfs/snapshot/hourly.12 zfs ro,relatime,xattr,noacl 0 0
datapool/Temp@hourly.1 /datapool/Temp/.zfs/snapshot/hourly.4 zfs ro,relatime,xattr,noacl 0 0
datapool/Temp@hourly.2 /datapool/Temp/.zfs/snapshot/hourly.5 zfs ro,relatime,xattr,noacl 0 0
datapool/Temp@hourly.0 /datapool/Temp/.zfs/snapshot/hourly.3 zfs ro,relatime,xattr,noacl 0 0
# umount /datapool/Temp/.zfs/snapshot/hourly.3
...
```

But I still don't know what happened there. And I still get those i/o errors, although I seem to be able to access the snapshots again. At least my script that finds file references in snapshots (i.e., that looks into each snapshot) is able to stat files within all the snapshots atm, without getting any of the two errors.

I've found a similar bug report - #6489 - but that one seems to be about send/receive. The system in question has zfsutils-linux from the Debian stable repository installed.

### Describe how to reproduce the problem

```
$ cd /datapool/Temp
$ ls -la .zfs/snapshot/ >/dev/null 
ls: .zfs/snapshot/: Input/output error
ls: .zfs/snapshot/.: Input/output error
ls: .zfs/snapshot/..: Input/output error
ls: .zfs/snapshot/hourly.33: Input/output error
ls: .zfs/snapshot/hourly.44: Input/output error
ls: .zfs/snapshot/hourly.39: Input/output error
ls: .zfs/snapshot/hourly.20: Input/output error
ls: .zfs/snapshot/hourly.1: Input/output error
ls: .zfs/snapshot/hourly.18: Input/output error
ls: .zfs/snapshot/hourly.12: Input/output error
ls: .zfs/snapshot/weekly.0: Input/output error
...
```

### Include any warning/errors/backtraces from the system logs

I don't see any errors in dmesg.

**Comment by cyphar** [comment]
Can you do an `strace ls` so it's clearer what syscalls are failing with `EIO`?

**Comment by c0xc** [comment]
The snapshot rotation is failing again, triggering cron emails:

```
cannot rename 'datapool/Temp@hourly.47': dataset is busy
Error: Renaming snapshot failed!
datapool/Temp@hourly.47
```

I'm getting the other error again:

```
# ls -ld .zfs/snapshot/hourly.47/
ls: cannot access '.zfs/snapshot/hourly.47/': Too many levels of symbolic links
```

I'm attaching a trace. Apparently, it can't find some locale files. What does that have to do with ZFS?

[trace1.txt](https://github.com/zfsonlinux/zfs/files/3211462/trace1.txt)

**Comment by cyphar** [comment]
The locale `ENOENT` errors can be ignored, that's normal. The key point of the three traces is:

```
% grep ELOOP trace1.txt
lstat(".zfs/snapshot/hourly.47/", 0x55bcee3ba1a0) = -1 ELOOP (Too many levels of symbolic links)
```
```
% grep EIO trace2.txt
lgetxattr(".zfs/snapshot/hourly.0", "security.selinux", 0x5558492727f0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.1", "security.selinux", 0x55584927b1d0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.10", "security.selinux", 0x55584927b230, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.11", "security.selinux", 0x55584927b250, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.12", "security.selinux", 0x55584927b270, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.13", "security.selinux", 0x55584927b290, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.14", "security.selinux", 0x55584927b2b0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.15", "security.selinux", 0x55584927b2d0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.16", "security.selinux", 0x55584927b2f0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.17", "security.selinux", 0x55584927b310, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.18", "security.selinux", 0x55584927b330, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.19", "security.selinux", 0x55584927b350, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.2", "security.selinux", 0x55584927b370, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.20", "security.selinux", 0x55584927b390, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.21", "security.selinux", 0x55584927b3b0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.22", "security.selinux", 0x55584927b3d0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.23", "security.selinux", 0x55584927b3f0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.24", "security.selinux", 0x55584927b410, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.25", "security.selinux", 0x55584927b430, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.26", "security.selinux", 0x55584927b450, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.27", "security.selinux", 0x55584927b470, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.28", "security.selinux", 0x55584927b490, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.29", "security.selinux", 0x55584927b4b0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.3", "security.selinux", 0x55584927b4d0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.30", "security.selinux", 0x55584927b4f0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.31", "security.selinux", 0x55584927b510, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.32", "security.selinux", 0x55584927b530, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.33", "security.selinux", 0x55584927b550, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.34", "security.selinux", 0x55584927b570, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.35", "security.selinux", 0x55584927b590, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.36", "security.selinux", 0x55584927b5b0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.37", "security.selinux", 0x55584927b5d0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.38", "security.selinux", 0x55584927b5f0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.39", "security.selinux", 0x55584927b610, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.4", "security.selinux", 0x55584927b630, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.40", "security.selinux", 0x55584927b650, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.41", "security.selinux", 0x55584927b670, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.42", "security.selinux", 0x55584927b690, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.43", "security.selinux", 0x55584927b6b0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.44", "security.selinux", 0x55584927b6d0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.45", "security.selinux", 0x55584927b6f0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.47", "security.selinux", 0x55584927b710, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.5", "security.selinux", 0x55584927b770, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.6", "security.selinux", 0x55584927b790, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.69", "security.selinux", 0x55584927b7b0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.7", "security.selinux", 0x55584927b7f0, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.70", "security.selinux", 0x55584927b810, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.8", "security.selinux", 0x55584927b830, 255) = -1 EIO (Input/output error)
lgetxattr(".zfs/snapshot/hourly.9", "security.selinux", 0x55584927b850, 255) = -1 EIO (Input/output error)
```
```
% grep EREMOTE trace3.txt
lstat("/datapool/Temp/.zfs/snapshot/hourly.49/AUTOCLEAN/", 0x55fff70c41a0) = -1 EREMOTE (Object is remote)
```

I suspected that the EIO would be for `xattrs` (mainly because you said you could access the files and so it seemed likely that the file contentss weren't the thing that were giving errors). I can't be of much more help, but hopefully the above might help some of the ZFS folks will have more debugging tips for missing xattrs in snapshots?

**Comment by c0xc** [comment]
Well, that reminds me: I have enabled acltype not too long ago. That was on another filesystem though (same pool). So that's probably unrelated.

I have to make a clarification: I did say that I was still able to access files in the snapshots, but that was when I got the i/o error message. The directory listing came out on stdout and that error came out on stderr. However: The symlink error (and the third error as well) as described in my last comment is currently preventing me from accessing files in the affected snapshot. Meanwhile, other snapshots of the same filesystem are accessible without causing any error at all.
